components/webext-storage/src/migration.rs

```diff
@@ -107,29 +108,42 @@ struct Parsed<'a> {
     data: serde_json::Value,
 }
 
-pub fn migrate(tx: &Transaction<'_>, filename: &Path) -> Result<usize> {
+pub fn migrate(tx: &Transaction<'_>, filename: &Path) -> Result<MigrationInfo> {
     // We do the grouping manually, collecting string values as we go.
     let mut last_ext_id = "".to_string();
     let mut curr_values: Vec<(String, serde_json::Value)> = Vec::new();
-    let mut num_extensions = 0;
-    for row in read_rows(filename) {
+    let mut curr_quota_bytes: usize = 0;
+    let (rows, mut mi) = read_rows(filename);
+    for row in rows {
         log::trace!("processing '{}' - '{}'", row.col_name, row.record);
         let parsed = match row.parse() {
             Some(p) => p,
-            None => continue,
+            None => {
+                mi.non_fatal_errors_seen += 1;
+                continue;
+            }
         };
         // Do our "grouping"
         if parsed.ext_id != last_ext_id {
             if last_ext_id != "" && !curr_values.is_empty() {
                 // a different extension id - write what we have to the DB.
-                do_insert(tx, &last_ext_id, curr_values)?;
-                num_extensions += 1;
+                let entries = do_insert(tx, &last_ext_id, curr_values)?;
+                mi.extensions_successful += 1;
+                mi.entries_successful += entries;
             }
             last_ext_id = parsed.ext_id.to_string();
             curr_values = Vec::new();
+            curr_quota_bytes = 0;
         }
         // no 'else' here - must also enter this block on ext_id change.
         if parsed.ext_id == last_ext_id {
+            let quota_bytes = crate::api::get_quota_size_of(&parsed.key, &parsed.data);
+            curr_quota_bytes += quota_bytes;
+            if quota_bytes > crate::api::SYNC_QUOTA_BYTES_PER_ITEM
+                || curr_quota_bytes > crate::api::SYNC_QUOTA_BYTES
+            {
+                mi.overflowed_quotas = true;
```

Comment:
I wonder if it's worth recording how much we overflowed the quota, using buckets and counts for how many times we overflowed in each one. (Say, 1-10 bytes, 11-20, 21-50, 51-100, 100-500, 501-1k, and 1k+, and adding one to the bucket when we spill over).

On the other hand, I don't know what we'll _do_ with that information. I'm imagining we could use it to bump the quota, and identify interesting trends—if most spillovers are 100 bytes or fewer, that's more actionable than many kilobytes. Buuuut, that would mean diverging from Chrome, and I suspect there's a long tail of spillovers that'll eventually bump up against the maximum Sync record limit, which we can't sync at all. And analyzing the data like that might be super involved...

I think `overflowed_quotas` is a good start, and bucketing overflows beyond that shouldn't block your PR. But just something I was thinking about while reviewing...

Reply:
This has been removed.